### PR TITLE
TLS: add `tls_handshake` hook, ignore-after-clienthello

### LINKED ISSUE #4702: Make it possible to enable TLS passthrough from `tls_clienthello` hook.
From Slack:

> given that I cannot decrypt the connection, I am  currently attempting to bypass the TLS connections altogether.
by doing the following:
```
def next_layer(self, nextlayer: layer.NextLayer):
    is_tls = is_tls_record_magic(nextlayer.data_client())
    if(is_tls):
        nextlayer.layer = layers.TCPLayer(nextlayer.context, ignore=True)
```
> This allows the device to function without breaking due to TLS cert errors, however if I do this, the tls_clienthello hook does not fire. Inversely, if it fires, I do get the SNI but the connection fails.

----

We should look into what it takes to add a `tls_clienthello.passthrough: bool` attribute, which when turned on, makes us not intercept the connection. I have an idea what to do here and will look into it.

### PR DESCRIPTION
**Pull request by mhils**
This PR adds two new features to our TLS implementation:

1. A new `tls_handshake` hook, which gets fired after a TLS handshake has been completed (successfully or not). I'm not sure yet if the name is perfect, feedback is welcome.
2. Make it possible to stop TLS interception from the `tls_clienthello` event hook.

Additionally, I've revised the `tls_passthrough.py` script as an example for both. Now without monkeypatching. 😄 

To do:

 - [x] code review
 - [x] tests
 - [x] add docs

fixes #4702

### CODE REVIEW

**Comment by Prinzhorn** [comment]
Neat, I don't have the mental juice right now to give it a full review though

1. Shouldn't there be a `_server` and a `_client` variant? There already is a hook with a `if isinstance(data.conn, connection.Server)` in this PR
2. This does fix #3884 as well, right? Or is that not the earliest possible hook to get the cert info?
3. Not sure how I feel about having the same hook for success and failure. But I don't know enough about handshakes and the use-cases of this hook to judge

File: docs/scripts/api-events.py [hunk 1]
```diff
@@ -124,6 +124,7 @@ def category(name: str, desc: str, hooks: List[Type[hooks.Hook]]) -> None:
             tls.TlsClienthelloHook,
             tls.TlsStartClientHook,
             tls.TlsStartServerHook,
+            tls.TlsHandshakeHook,
```
Comment (Kriechi) [inline docs/scripts/api-events.py:127]: Any proposals how to make the naming somehow include the "completed" idea of this hook?
`TlsHandshakeCompletedHook` or similar? Or using the `...Start/End...` scheme?

**Comment by Kriechi** [comment]
Looks great! Only nitpicking the event name a bit 😄

**Comment by 89z** [comment]
https://github.com/mitmproxy/mitmproxy/issues/4877

**Comment by mhils** [comment]
Summarizing the naming feedback:

> @Kriechi: Any proposals how to make the naming somehow include the "completed" idea of this hook? `TlsHandshakeCompletedHook` or similar? Or using the `...Start/End...` scheme?

The problem I'm having with _end_ is that the TLS connection is not over at this point, only the handshake has completed.

> @Prinzhorn: Shouldn't there be a `_server` and a `_client` variant? There already is a hook with a `if isinstance(data.conn, connection.Server)` in this PR

Yes, that makes sense. 👍

> @Prinzhorn: Not sure how I feel about having the same hook for success and failure. But I don't know enough about handshakes and the use-cases of this hook to judge. If it does solve 2. for me I'd only like successes, but sure I can check for those

Thank you for explicitly taking the API consumer point-of-view, that is good. Makes sense to me. 👍

How would you all feel about splitting `tls_handshake` into the following four event names:

- `tls_established_client`
- `tls_established_server`
- `tls_error_client`
- `tls_error_server`

**Comment by Kriechi** [comment]
I like the `_established_` part! 

I know the `_error_` part follows existing patterns, though I feel like it doesn't match "language wise", a past-tense adjective.  How about `_established_` + `_failed_`?

**Comment by mhils** [comment]
I have no strong preference. Let's invoke democracy:

- React with 🎉  to this comment if you prefer `tls_error_client`.
- React with 🚀 to this comment if you prefer `tls_failed_client`.

**Comment by Prinzhorn** [comment]
I have `github.com##.js-reactions-container` in my uBlock Origin filters, so much for accessibility to your corrupt voting system :smile: 

> I have no strong preference.

Same, with a tendency for :rocket: because consistency within the `tls_*` category at least. Or what will happen with the existing three `tls_` hooks? They're all nouns, `tls_start_client` vs `tls_started_client` :sweat_smile:

**Comment by mhils** [comment]
Available on main now: https://docs.mitmproxy.org/dev/api/events.html#TLSEvents.tls_established_client 😃

**Comment by 89z** [comment]
?

**Comment by mhils** [comment]
https://docs.mitmproxy.org/dev/api/events.html#TLSEvents.tls_clienthello has all the relevant bits - sorry for the slowish reply, merry christmas!

**Comment by 89z** [comment]
Thanks for the info, but I dont know how that link works with the command line tools...

**Comment by mhils** [comment]
You can use the addon API with all mitmproxy tools:
https://docs.mitmproxy.org/dev/addons-overview/
The previous link points to the `tls_clienthello` hook and describes the argument it receives.

**Pull request merged**